Revised FEI stream reader, see Issue #2385

din14970 opened the pull request:
### Description of the change
A few months ago on my other computer, reading of FEI Velox emd was excruciatingly slow with Hyperspy and was not compatible with files produced in older Velox versions. Couldn't really figure out the issue so I started writing my own functions to read the format with `h5py`. I haven't been able to reproduce these issues in another install on another computer, which is kind  of frustrating - my emd files now read fast and the old ones now also work. **So now I'm not 100% sure my contribution is actually an improvement in terms of performance. But it does make the code look cleaner.**

- Main thing I changed was the `misc/io/fei_stream_readers`. This file had 4 different functions that converted the spectrum stream into a sparse or full matrix. I condensed this into one vectorized one `_stream_to_scipy_sparse_COO` and edited the functions that call the 4 previous versions. The function is a modification of what I describe in the issue to make it compatible with the various arguments. The function does rely on the entire `stream` matrix to be read into memory. 
  - I added the `jit_ifnumba` decorator but I'm not sure whether this will work on my implementation
  - So far I was unable to run the tests as I ran into  #2402 which prevented me from running the tests locally. I did some informal tests with the code copied into a jupyter notebook and there it seems to work as expected.

- added one line to .gitignore to ignore my virtual environment

### Progress of the PR

### Minimal example of the bug fix or the new feature
No new feature or bug fix, code just looks cleaner and 4 functions are condensed into 1

din14970 committed:
Vectorized and simplified fei stream reader

din14970 committed:
ignore venv

din14970 commented:
Ok so `jit_ifnumba` doesn't seem to work and makes all the tests fail

francisco-dlp commented:
Just remove it a let's see how it goes. If the tests pass, it'll be interesting to benchmark both approaches.

ericpre commented:
> * So far I was unable to run the tests as I ran into  #2402 which prevented me from running the tests locally. I did some informal tests with the code copied into a jupyter notebook and there it seems to work as expected.

This is fixed in the development branch (`RELEASE_next_minor` for example): merging your branch with upstream will fix it.

din14970 committed:
bug fix to match tests, uncertainty about pixel offset

din14970 commented:
I branched from `RELEASE_next_minor` and pulled changes yesterday, then installed tests in `./.venv` with `pip install -e .[test]`, maybe did something wrong? Anyway, downgrading `scikit-image` to `0.16.2` seems to do the trick and I can run at least the emd specific tests on my machine.

Managed to fix the issue with `jit_ifnumba`, problem was that i put it above a function that returned arrays of variable shapes. I split up the function into summed versus not summed and put the decorator above. Not sure whether it is doing anything though, but at least it doesn't throw errors.

By being able to run the tests, I encountered other issues, noteably in the tests, stemming from the function `array_to_stream` and ultimately the interpretation of the stream. This leads me to believe that the original functions contain a small bug resulting in a one-pixel offset of the entire spectrum image. Reasoning:

From the real streams in emd files I have seen, the number of times 65535 appears is exactly equal to `(xdim * ydim * frames)`, except if the last collection frame was partial. All streams I have seen start with 65535. Hence, I interpret this as the first scan position (frame, y, x) = (0,0,0), with the channel numbers coming after this mapped as counts to this position. `array_to_stream` produces a stream that does not start with 65535 and has `(xdim*ydim*frames)-1` occurences of 65535. After a small edit to `array_to_stream`, forcing it to start with 65535 I could get the tests in `tests/misc/test_fei_stream_readers.py` to pass with my implementation.

The original functions do it the other way around - counts are mapped to the next occurence of 65535. Due to the observation on the 65535 count, I think this is incorrect.

Therefore, the tests in `tests/io/test_emd.py` still failed because of the comparison with the `.npy` file, which due to this one-pixel offset does not match with my spectrum stream. They key is in the line (in my version):

```
pixel_indexes = count_channel_regs - np.arange(count_channel_regs.shape[0]) - 1
```
If I change the -1 to 0, I can get the tests in `tests/io/test_emd.py` to pass. However, I am reasonably confident that the npy file is wrong and that there should be a -1 there.

I have now modified the code so that this -1 offset is NOT taken into account so that all tests should pass. However, I am convinced instead the npy file should be updated.

What is your opinion?

ericpre commented:
> I branched from `RELEASE_next_minor` and pulled changes yesterday, then installed tests in `./.venv` with `pip install -e .[test]`, maybe did something wrong? Anyway, downgrading `scikit-image` to `0.16.2` seems to do the trick and I can run at least the emd specific tests on my machine.

If you have pulled your branch from your fork, it is very likely that it was not up to date:
https://github.com/din14970/hyperspy/commits/RELEASE_next_minor - the branch of this PR is based on https://github.com/din14970/hyperspy/commit/677c15a3585f3e540c9a78a22fb02f7b7ea0959d.

numba is an optional dependencies and installing hyperspy using `pip install -e .[tests]` will not install numba, while `pip install -e .[dev]` will install all dependencies. The latter has been added since the last release as a convenience and is documented in the development branch but it is not part of the release documentation yet.

As mentioned by @francisco-dlp, it would be good to benchmark the difference versions:
- before this PR,
- with this PR and numba,
- with this PR without numba.

Since It seems almost fully vectorized, it is fairly likely that numba will not improve (and possibly decrease) the performance!

Thanks for the detailed explanation, I will have a look at it!

din14970 commented:
> If you have pulled your branch from your fork, it is very likely that it was not up to date:
https://github.com/din14970/hyperspy/commits/RELEASE_next_minor - the branch of this PR is based on din14970@677c15a.

Right, of course. Sorry, I'm just an amateur.

>  As mentioned by @francisco-dlp, it would be good to benchmark the difference versions

Never looked into this before, so not sure if this is the proper way to do it. I used [pytest-benchmark](https://pytest-benchmark.readthedocs.io/en/latest/index.html) and created a couple extra bench mark tests in a separate file for each combination based on the existing ones in `test_fei_stream_readers`. Then I put all the different implementations back into the `fei_stream_readers` file and gave them separate names. Results:

```
----------------------------------------------------------------------------------------------------------- benchmark: 12 tests ------------------------------------------------------------------------------------------------------------
Name (time in ms)                                                      Min                   Max                Mean              StdDev              Median                 IQR            Outliers       OPS            Rounds  Iterations
--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
test_benchmark_stream_to_sparse_COO_array[False]                    6.0586 (1.0)          6.5793 (1.0)        6.3545 (1.00)       0.2209 (1.0)        6.3738 (1.02)       0.3798 (1.16)          2;0  157.3684 (1.00)          5           1
test_benchmark_stream_to_sparse_COO_array[True]                     6.1345 (1.01)         6.8007 (1.03)       6.3512 (1.0)        0.2706 (1.22)       6.2317 (1.0)        0.3277 (1.0)           1;0  157.4499 (1.0)           5           1
test_benchmark_stream_to_array[True]                               46.5074 (7.68)        48.1856 (7.32)      47.2697 (7.44)       0.7205 (3.26)      46.9071 (7.53)       1.1720 (3.58)          2;0   21.1552 (0.13)          5           1
test_benchmark_stream_to_array[False]                              48.2451 (7.96)        50.0751 (7.61)      48.8431 (7.69)       0.7282 (3.30)      48.4982 (7.78)       0.7663 (2.34)          1;0   20.4737 (0.13)          5           1
test_benchmark_stream_to_array_vector_numba[False]                295.6466 (48.80)      301.5567 (45.83)    298.5121 (47.00)      2.5499 (11.54)    299.4577 (48.05)      4.2840 (13.07)         2;0    3.3499 (0.02)          5           1
test_benchmark_stream_to_sparse_COO_array_vector_numba[False]     297.2040 (49.05)      319.3619 (48.54)    310.5124 (48.89)      9.8387 (44.54)    315.8068 (50.68)     16.3246 (49.81)         1;0    3.2205 (0.02)          5           1
test_benchmark_stream_to_array_vector_numba[True]                 332.2100 (54.83)      385.7956 (58.64)    365.1151 (57.49)     23.7133 (107.35)   378.2802 (60.70)     38.6100 (117.81)        1;0    2.7389 (0.02)          5           1
test_benchmark_stream_to_sparse_COO_array_vector_numba[True]      335.9396 (55.45)      356.1144 (54.13)    345.0229 (54.32)      8.1677 (36.98)    344.4114 (55.27)     13.2575 (40.45)         2;0    2.8984 (0.02)          5           1
test_benchmark_stream_to_array_vector[True]                       409.6479 (67.61)      414.4877 (63.00)    411.9575 (64.86)      1.7860 (8.09)     411.5109 (66.03)      2.1572 (6.58)          2;0    2.4274 (0.02)          5           1
test_benchmark_stream_to_sparse_COO_array_vector[True]            421.6352 (69.59)    1,125.1943 (171.02)   600.6156 (94.57)    303.2999 (>1000.0)  429.6974 (68.95)    311.2097 (949.56)        1;0    1.6650 (0.01)          5           1
test_benchmark_stream_to_sparse_COO_array_vector[False]           428.5342 (70.73)      451.5604 (68.63)    435.1953 (68.52)      9.3571 (42.36)    431.4093 (69.23)      8.3947 (25.61)         1;1    2.2978 (0.01)          5           1
test_benchmark_stream_to_array_vector[False]                      430.2905 (71.02)      955.0198 (145.16)   558.8440 (87.99)    222.7775 (>1000.0)  476.6715 (76.49)    166.1984 (507.10)        1;1    1.7894 (0.01)          5           1
--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
```

The True/False is whether the frames are summed or not. So my conclusion would be that the original implementation is the fastest by far, especially the COO one, then my implementation + numba which is about 6x slower, then my implementation without numba which is about 8x slower. Somewhat unfortunate because I thought my method looked quite elegant and did the same thing with much fewer lines. Not sure exactly where the slow-ness originates, maybe should time the different parts of the functions to optimize but have no time at the moment. I have pushed this commit onto the branch `benchmarking` in my fork if anyone wants to check it out.

Still, I believe the functions should be corrected for the fact that all spectrumstreams start with 65535; the 1 pixel offset mistake in the readers means that the npy file for comparison in the tests is wrong.

din14970 commented:
In the comment above it was run on the dense array, here are the results for the sparse array:

```
------------------------------------------------------------------------------------------------------------- benchmark: 12 tests --------------------------------------------------------------------------------------------------------------
Name (time in us)                                                      Min                   Max                Mean              StdDev              Median                 IQR            Outliers  OPS (Kops/s)            Rounds  Iterations
------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
test_benchmark_stream_to_array[False]                               2.6270 (1.0)         24.3260 (1.48)       5.9441 (1.12)       8.1176 (2.13)       2.6510 (1.0)        0.9738 (1.0)           1;1      168.2328 (0.89)          7           1
test_benchmark_stream_to_array[True]                                2.6600 (1.01)        16.3930 (1.0)        5.3013 (1.0)        5.4629 (1.43)       2.9370 (1.11)       1.4650 (1.50)          1;1      188.6318 (1.0)           6           1
test_benchmark_stream_to_array_vector_numba[True]                 117.1790 (44.61)      137.7170 (8.40)     124.0888 (23.41)      8.0753 (2.12)     122.6390 (46.26)      8.7143 (8.95)          1;0        8.0587 (0.04)          5           1
test_benchmark_stream_to_array_vector_numba[False]                120.0910 (45.71)      267.0950 (16.29)    166.5762 (31.42)     59.4559 (15.59)    142.6740 (53.82)     67.6713 (69.50)         1;0        6.0033 (0.03)          5           1
test_benchmark_stream_to_array_vector[True]                       137.4930 (52.34)    1,404.7840 (85.69)    186.5830 (35.20)    110.5271 (28.98)    145.1965 (54.77)     55.3325 (56.82)       23;27        5.3595 (0.03)        280           1
test_benchmark_stream_to_array_vector[False]                      140.7110 (53.56)    2,307.3250 (140.75)   189.7436 (35.79)    113.5509 (29.78)    157.0180 (59.23)     32.8875 (33.77)     252;549        5.2703 (0.03)       3475           1
test_benchmark_stream_to_sparse_COO_array[True]                   539.9070 (205.52)     739.1190 (45.09)    609.9848 (115.06)    81.7856 (21.45)    591.2420 (223.03)   116.9280 (120.08)        1;0        1.6394 (0.01)          5           1
test_benchmark_stream_to_sparse_COO_array_vector_numba[True]      569.6740 (216.85)   5,129.2280 (312.89)   759.2133 (143.21)   453.7616 (118.99)   588.5445 (222.01)   148.3735 (152.37)     91;166        1.3172 (0.01)       1280           1
test_benchmark_stream_to_sparse_COO_array[False]                  572.1390 (217.79)     582.3000 (35.52)    577.2374 (108.89)     3.8133 (1.0)      577.5670 (217.87)     5.2215 (5.36)          2;0        1.7324 (0.01)          5           1
test_benchmark_stream_to_sparse_COO_array_vector_numba[False]     601.5480 (228.99)   3,508.6760 (214.04)   741.9386 (139.95)   354.6941 (93.01)    612.7420 (231.14)    77.7740 (79.87)      97;221        1.3478 (0.01)       1308           1
test_benchmark_stream_to_sparse_COO_array_vector[True]            632.3100 (240.70)     887.0690 (54.11)    688.6694 (129.90)   111.2560 (29.18)    635.9800 (239.90)    78.8712 (81.00)         1;1        1.4521 (0.01)          5           1
test_benchmark_stream_to_sparse_COO_array_vector[False]           641.6900 (244.27)   2,162.7340 (131.93)   715.5740 (134.98)   140.0418 (36.72)    670.1255 (252.78)    49.1910 (50.52)      95;152        1.3975 (0.01)       1182           1
------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
```

ericpre commented:
Thanks for the benchmark, this is useful!

> The True/False is whether the frames are summed or not. So my conclusion would be that the original implementation is the fastest by far, especially the COO one, then my implementation + numba which is about 6x slower, then my implementation without numba which is about 8x slower. Somewhat unfortunate because I thought my method looked quite elegant and did the same thing with much fewer lines. Not sure exactly where the slow-ness originates, maybe should time the different parts of the functions to optimize but have no time at the moment. I have pushed this commit onto the branch `benchmarking` in my fork if anyone wants to check it out.

Indeed, may be worth revisiting a some point. I would suggest to proceed in the following way:
- open a new PR to sort out the issue you mentioned with the pixel offset
- close this PR so that we keep a record of your work and what you have found - we can come back to it later to understand what is going on.

tjof2 commented on hyperspy/misc/io/fei_stream_readers.py:
```diff
@@ -219,77 +55,147 @@ def stream_to_sparse_COO_array(
+    """
+    frms, ydim, xdim, chnx = shape
+    # the indexes where counts are registered
+    count_channel_regs = np.argwhere(stream != 65535)[:, 0]
```
What shape is `stream`?

Is it 1D? If so, what does `np.flatnonzero(stream != 65535)` do instead of the `argwhere`?

If it's not 1D, then these are equivalent I think:
```python
np.argwhere(stream != 65535)[:, 0]
np.unravel_index(np.flatnonzero(stream != 65535), stream.shape).T[:, 0]
```

din14970 replied:
Tried it out, cool function that I didn't know about. No significant speed-up unfortunately neither with a sparse nor full matrix.

tjof2 reviewed (commented):
I wonder if the benchmark is poor because of `np.argwhere`

din14970 commented:
> This leads me to believe that the original functions contain a small bug resulting in a one-pixel offset of the entire spectrum image. Reasoning:
> 
> From the real streams in emd files I have seen, the number of times 65535 appears is exactly equal to `(xdim * ydim * frames)`, except if the last collection frame was partial. All streams I have seen start with 65535. Hence, I interpret this as the first scan position (frame, y, x) = (0,0,0), with the channel numbers coming after this mapped as counts to this position. `array_to_stream` produces a stream that does not start with 65535 and has `(xdim*ydim*frames)-1` occurences of 65535. After a small edit to `array_to_stream`, forcing it to start with 65535 I could get the tests in `tests/misc/test_fei_stream_readers.py` to pass with my implementation.
> 
> The original functions do it the other way around - counts are mapped to the next occurence of 65535. Due to the observation on the 65535 count, I think this is incorrect.

By chance we got a dataset where the first entry in the stream was indeed not 65535, where my reader failed. Just wanted to report this to say that I was wrong and a bug fix will not be needed. I thought it made so much sense since there are exactly `(xdim * ydim * frames)` times 65535. Sorry for doubting you, hyperspy developers ;)

ericpre commented:
Thanks @din14970 for coming back on this. It may be valuable information for future improvement!
